### PR TITLE
N°4692 - Enable parallelization of multiple CRON jobs

### PR DESCRIPTION
**Pull request by eespie**
Allow cron process to execute in parallel running different tasks.

Each task is only executed once at a time, but different tasks can run in parallel if multiple cron are executed.
The maximum number of parallel cron processes is configurable in iTop configuration.

The number of concurrent cron processes is obtained by configuring the 'cron_max_execution_time' and playing with the period of the cron. For example if the cron process last 10 minutes and the cron is launched every minute there will be roughly 10 parallel  cron running.

Another change is that the 'next_run_date' is calculated before running the task in order to have an update in case of problem and avoid running the same task again and again.

### CODE REVIEW

File: webservices/cron.php [hunk 1]
```diff
@@ -320,12 +318,12 @@ function CronExec($oP, $bVerbose, $bDebug=false)
 				}
 			}
 		}
-
-		if ($bVerbose && $bWorkDone)
-		{
-			$oP->p("Sleeping...\n");
+		if ($aTasks == []) {
```
Comment (Hipska) [inline webservices/cron.php]: Wouldn't it be better to do this?
```suggestion
		if (empty($aTasks)) {
```

**Review by Hipska — APPROVED** [review]
Looks good. Plain simple implementation!

File: core/config.class.inc.php [hunk 2]
```diff
@@ -512,6 +504,14 @@ class Config
 			'source_of_value' => '',
 			'show_in_conf_sample' => false,
 		],
+		'cron.max_process' => [
```
Comment (Molkobain) [inline core/config.class.inc.php]:
```suggestion
		'cron.max_processes' => [
```

File: core/config.class.inc.php [hunk 3]
```diff
@@ -512,6 +504,14 @@ class Config
...
+			'type' => 'integer',
+			'description' => 'Maximum number of cron process to run',
```
Comment (Molkobain) [inline core/config.class.inc.php]:
```suggestion
			'description' => 'Maximum number of cron processes to run',
```

File: setup/setuputils.class.inc.php [hunk 4]
```diff
@@ -2033,30 +2033,40 @@ private static function WaitCronTermination($oConfig, $sMode)
 			if (is_null($oConfig) || ContextTag::Check(ContextTag::TAG_CRON)) {
 				return;
 			}
-			// Use mutex to check if cron is running
+			// Limit the number of cron process to run in parallel
+			$iMaxCronProcess = $oConfig->Get('cron.max_process');
```
Comment (Molkobain) [inline setup/setuputils.class.inc.php]:
```suggestion
			$iMaxCronProcess = $oConfig->Get('cron.max_processes');
```

File: webservices/cron.php [hunk 5]
```diff
@@ -209,6 +187,7 @@ function CronExec($oP, $bVerbose, $bDebug=false)
 	$iMaxDuration = MetaModel::GetConfig()->Get('cron_max_execution_time');
 	$iTimeLimit = $iStarted + $iMaxDuration;
 	$iCronSleep = MetaModel::GetConfig()->Get('cron_sleep');
+	$iMaxCronProcess = MetaModel::GetConfig()->Get('cron.max_process');
```
Comment (Molkobain) [inline webservices/cron.php]:
```suggestion
	$iMaxCronProcess = MetaModel::GetConfig()->Get('cron.max_processes');
```

File: webservices/cron.php (hunk 1 above)
Comment (Molkobain) [reply webservices/cron.php]: Yes, but we rather use `count($aTasks) === 0` than `empty()`

File: webservices/cron.php [hunk 6]
```diff
@@ -301,10 +295,14 @@ function CronExec($oP, $bVerbose, $bDebug=false)
...
 				CheckMaintenanceMode($oP);
+				if ($iMaxCronProcess > 1) {
+					// Reindex tasks every time
+					break;
+				}
 			}
 
 			// Tasks to run later
-			if ($bVerbose)
+			if ($bVerbose && $aTasks == [])
```
Comment (Molkobain) [inline webservices/cron.php]: There also

**Comment by Molkobain** [comment]
@eespie will:
- Create the corresponding bug in the bug tracker.
- Discuss with @Rorcha07 about it
- Study debug / verbose output when several CRON instances are running, which could be an issue when trying to debug something

File: sources/Service/Cron/CronLog.php [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+/*
+ * @copyright   Copyright (C) 2010-2022 Combodo SARL
+ * @license     http://opensource.org/licenses/AGPL-3.0
+ */
+
+namespace Combodo\iTop\Service\Cron;
+
+use LogAPI;
+
+class CronLog extends LogAPI
```
Comment (Molkobain) [inline sources/Service/Cron/CronLog.php:17]:
```suggestion
/**
 * @since 3.1.0
 */
class CronLog extends LogAPI
```

File: webservices/cron.php [hunk 8]
```diff
@@ -228,17 +217,10 @@ function CronExec($oP, $bVerbose, $bDebug=false)
...
-					$oP->p(sprintf('| %1$-25.25s | %2$-7s | %3$-19s | %4$-19s |', $sTaskName, $sStatus, $sLastRunDate, $sNextRunDate));
-				}
-			}
-			if ($bVerbose)
-			{
-				$oP->p('+---------------------------+---------+---------------------+---------------------+');
+				$sStatus = $oTask->Get('status');
+				$sLastRunDate = $oTask->Get('latest_run_date');
+				$sNextRunDate = $oTask->Get('next_run_date');
+				CronLog::Debug(sprintf('Task Class: %1$-25.25s Status: %2$-7s Last Run: %3$-19s Next Run: %4$-19s', $sTaskName, $sStatus, $sLastRunDate, $sNextRunDate));
```
Comment (Molkobain) [inline webservices/cron.php]: Can you give an example of output? Seems more complicated to read than the previous way as there will be no columns anymore.

File: webservices/cron.php [hunk 9]
```diff
@@ -313,33 +287,24 @@ function CronExec($oP, $bVerbose, $bDebug=false)
...
+						CronLog::Debug(sprintf("-- Skipping task: %-'-40s", $oTask->Get('class_name').' ')." until: ".$oTask->Get('next_run_date'));
 					}
 				}
 			}
 		}
-		if ($aTasks == []) {
-			if ($bVerbose) {
-				$oP->p("Sleeping...\n");
-			}
+		if (count($aTasks) == 0) {
```
Comment (Molkobain) [inline webservices/cron.php:296]: Missing "Sleeping...\n" debug message

File: webservices/cron.php [hunk 10]
```diff
@@ -569,40 +524,42 @@ function ReSyncProcesses($oP, $bVerbose, $bDebug)
...
 			}
 		}
 		if ($bCanRun) {
-			CronExec($oP, $bVerbose, $bDebug);
+			CronLog::$iProcessNumber = $iProcessNumber;
+			CronLog::Info('Starting: '.time().' ('.date('Y-m-d H:i:s').')');
+			CronExec($bDebug);
 		} else {
-			$oP->p("Already $iMaxCronProcess are running...");
+			CronLog::Debug("Already $iMaxCronProcess are running...");
```
Comment (Molkobain) [inline webservices/cron.php]: Having "Already 5 are running..." in the log might not be clear to understand for someone troubleshooting/browsing the logs if it doesn't know the code.

What do you think of "Already 5/5 CRON processes are running, cannot start another one..." (or something similar) ?

File: webservices/cron.php [hunk 11]
```diff
@@ -187,10 +167,6 @@ function RunTask(BackgroundTask $oTask, $iTimeLimit)
 }
 
 /**
- * @param CLIPage|WebPage $oP
- * @param boolean $bVerbose
- *
- * @param bool $bDebug
```
Comment (Hipska) [inline webservices/cron.php:174]: debug param is still present..

File: webservices/cron.php [hunk 12]
```diff
@@ -203,110 +179,106 @@ function RunTask(BackgroundTask $oTask, $iTimeLimit)
...
-			}
-			if ($bVerbose)
-			{
-				$oP->p('+---------------------------+---------+---------------------+---------------------+');
+				$aTasks[] = $oTask;
+				$sStatus = $oTask->Get('status');
+				$sLastRunDate = $oTask->Get('latest_run_date');
+				$sNextRunDate = $oTask->Get('next_run_date');
+				CronLog::Debug(sprintf('Task Class: %1$-25.25s Status: %2$-7s Last Run: %3$-19s Next Run: %4$-19s', $sTaskName, $sStatus, $sLastRunDate, $sNextRunDate));
 			}
```
Comment (Hipska) [inline webservices/cron.php]: I also think we have a mismatch with number of tasks planned and the number of tasks displayed. (When a tasks unable to get a mutex lock.)

File: webservices/cron.php [hunk 13]
```diff
@@ -203,110 +179,106 @@ function RunTask(BackgroundTask $oTask, $iTimeLimit)
...
 			$aRunTasks = [];
-			foreach ($aTasks as $oTask)
-			{
+			while ($aTasks != []) {
+				$oTask = array_shift($aTasks);
```
Comment (Hipska) [inline webservices/cron.php]: Why not a simple foreach? Also, might be better to use `count($aTasks)`.

Edit: Just saw that was originally the case..

**Review by Hipska — CHANGES_REQUESTED** [review]

File: core/config.class.inc.php [hunk 14]
```diff
@@ -496,14 +496,6 @@ class Config
...
 			'show_in_conf_sample' => true,
 		],
-		'cron_task_max_execution_time' => [
-			'type' => 'integer',
-			'description' => 'Background tasks will use this value (integer) multiplicated by its periodicity (in seconds) as max duration per cron execution. 0 is unlimited time',
-			'default' => 0,
-			'value' => 0,
-			'source_of_value' => '',
-			'show_in_conf_sample' => false,
-		],
```
Comment (Hipska) [inline core/config.class.inc.php:593]: `cron_task_max_execution_time` is a setting that should stay I assume?

**Comment by eespie** [comment]
Well, the parameter will stay, but the behaviour will change (the « multiplicated by its periodicity (in seconds) » part is not pertinent anymore)

File: core/config.class.inc.php (hunk 14 above)
Comment (eespie) [reply core/config.class.inc.php:593]: Well, the parameter will stay, but the behaviour will change (the « multiplicated by its periodicity (in seconds) » part is not pertinent anymore)

Comment (Hipska) [reply core/config.class.inc.php:593]: Okay, but the current state of this PR just removes this setting parameter. Or did I miss something?

Comment (eespie) [reply core/config.class.inc.php:506]: right, it's an error, I should have just change the description... thx

Comment (eespie) [reply core/config.class.inc.php:593]: In fact 'cron_max_execution_time' is yet used, but 'cron_task_max_execution_time' is not used anymore... sorry for the confusion. The limit of execution of one task is the remaining cron execution time as it won't block other tasks anymore.
I'm sorry this project is pretty old now, and I'm not sure it will be integrated soon :(

File: webservices/cron.php (hunk 8 above)
Comment (eespie) [reply webservices/cron.php]: The debug array has been removed because when logs are interleaved it's more complicated to read.
Here is a run of one cron instance:

```
cron1  3 Tasks planned to run now (2022-11-18 23:02:29):
cron1  Task Class: InlineImageGC             Status: active  Last Run: 2022-11-18 23:02:27 Next Run: 2022-11-18 23:02:28
cron1  Task Class: ExecAsyncTask             Status: active  Last Run: 2022-11-18 23:02:27 Next Run: 2022-11-18 23:02:29
cron1  Task Class: PersonalDataAnonymizer    Status: active  Last Run: 2022-11-18 23:02:29 Next Run: 2022-11-18 23:02:29
cron1  >> === 2022-11-18 23:02:29 Start task: InlineImageGC ==================================
cron1  InlineImageGC: Cleaned 0 old temporary InlineImage(s) and 0 old temporary Attachment(s).
cron1  << === 2022-11-18 23:02:29 End of:     InlineImageGC ================================== Next: 2022-11-18 23:02:30
cron1  2 Tasks planned to run now (2022-11-18 23:02:29):
cron1  Task Class: ExecAsyncTask             Status: active  Last Run: 2022-11-18 23:02:27 Next Run: 2022-11-18 23:02:29
cron1  Task Class: PersonalDataAnonymizer    Status: active  Last Run: 2022-11-18 23:02:29 Next Run: 2022-11-18 23:02:29
cron1  >> === 2022-11-18 23:02:29 Start task: ExecAsyncTask ==================================
cron1  ExecAsyncTask: processed 0 tasks
cron1  << === 2022-11-18 23:02:29 End of:     ExecAsyncTask ================================== Next: 2022-11-18 23:02:31
cron1  1 Tasks planned to run now (2022-11-18 23:02:29):
cron1  Task Class: PersonalDataAnonymizer    Status: active  Last Run: 2022-11-18 23:02:29 Next Run: 2022-11-18 23:02:29
cron1  >> === 2022-11-18 23:02:29 Start task: PersonalDataAnonymizer =========================
cron1  << === 2022-11-18 23:02:29 End of:     PersonalDataAnonymizer ========================= Next: 2022-11-18 23:02:29
cron1  -- Skipping task: InlineImageGC -------------------------- until: 2022-11-18 23:02:30
cron1  -- Skipping task: ExecAsyncTask -------------------------- until: 2022-11-18 23:02:31
cron1  -- Skipping task: CheckStopWatchThresholds --------------- until: 2022-11-18 23:02:33
cron1  -- Skipping task: FullTextBackgroundRefresh -------------- until: 2022-11-18 23:02:43
cron1  -- Skipping task: ObsolescenceDateUpdater ---------------- until: 2022-11-18 23:12:07
cron1  -- Skipping task: BulkExportResultGC --------------------- until: 2022-11-19 11:12:51
cron1  -- Skipping task: FullTextBackgroundRebuildFullIndex ----- until: 2022-11-20 01:30:00
cron1  -- Skipping task: LogFileRotationProcess ----------------- until: 2022-12-01 00:00:00
cron1  -- Skipping task: BackupExec ----------------------------- until: 3000-01-01 00:00:00
cron1  sleeping...
```

Comment (eespie) [reply webservices/cron.php]: An interleaved example (yes it's hard to follow):

```
2022-11-18 23:41:23 | Debug   | 1     | cron1  4 Tasks planned to run now (2022-11-18 23:41:23): | Cron |||
2022-11-18 23:41:23 | Debug   | 1     | cron2  4 Tasks planned to run now (2022-11-18 23:41:23): | Cron |||
2022-11-18 23:41:23 | Debug   | 1     | cron1  Task Class: PersonalDataAnonymizer    Status: active  Last Run: 2022-11-18 23:41:21 Next Run: 2022-11-18 23:41:21 | Cron |||
2022-11-18 23:41:23 | Debug   | 1     | cron2  Task Class: PersonalDataAnonymizer    Status: active  Last Run: 2022-11-18 23:41:21 Next Run: 2022-11-18 23:41:21 | Cron |||
2022-11-18 23:41:23 | Debug   | 1     | cron1  Task Class: InlineImageGC             Status: active  Last Run: 2022-11-18 23:41:21 Next Run: 2022-11-18 23:41:22 | Cron |||
2022-11-18 23:41:23 | Debug   | 1     | cron2  Task Class: InlineImageGC             Status: active  Last Run: 2022-11-18 23:41:21 Next Run: 2022-11-18 23:41:22 | Cron |||
2022-11-18 23:41:23 | Debug   | 1     | cron1  Task Class: ExecAsyncTask             Status: active  Last Run: 2022-11-18 23:41:21 Next Run: 2022-11-18 23:41:23 | Cron |||
2022-11-18 23:41:23 | Debug   | 1     | cron2  Task Class: ExecAsyncTask             Status: active  Last Run: 2022-11-18 23:41:21 Next Run: 2022-11-18 23:41:23 | Cron |||
2022-11-18 23:41:23 | Debug   | 1     | cron1  Task Class: CheckStopWatchThresholds  Status: active  Last Run: 2022-11-18 23:41:13 Next Run: 2022-11-18 23:41:23 | Cron |||
2022-11-18 23:41:23 | Debug   | 1     | cron2  Task Class: CheckStopWatchThresholds  Status: active  Last Run: 2022-11-18 23:41:13 Next Run: 2022-11-18 23:41:23 | Cron |||
2022-11-18 23:41:23 | Debug   | 1     | cron2  >> === 2022-11-18 23:41:23 Start task: PersonalDataAnonymizer ========================= | Cron |||
2022-11-18 23:41:23 | Debug   | 1     | cron1  >> === 2022-11-18 23:41:23 Start task: InlineImageGC ================================== | Cron |||
2022-11-18 23:41:23 | Debug   | 1     | cron1  InlineImageGC: Cleaned 0 old temporary InlineImage(s) and 0 old temporary Attachment(s). | Cron |||
2022-11-18 23:41:23 | Debug   | 1     | cron1  << === 2022-11-18 23:41:23 End of:     InlineImageGC ================================== Next: 2022-11-18 23:41:24 | Cron |||
2022-11-18 23:41:23 | Debug   | 1     | cron2  << === 2022-11-18 23:41:23 End of:     PersonalDataAnonymizer ========================= Next: 2022-11-18 23:41:23 | Cron |||
2022-11-18 23:41:23 | Debug   | 1     | cron1  3 Tasks planned to run now (2022-11-18 23:41:23): | Cron |||
2022-11-18 23:41:23 | Debug   | 1     | cron1  Task Class: ExecAsyncTask             Status: active  Last Run: 2022-11-18 23:41:21 Next Run: 2022-11-18 23:41:23 | Cron |||
2022-11-18 23:41:23 | Debug   | 1     | cron1  Task Class: CheckStopWatchThresholds  Status: active  Last Run: 2022-11-18 23:41:13 Next Run: 2022-11-18 23:41:23 | Cron |||
2022-11-18 23:41:23 | Debug   | 1     | cron1  Task Class: PersonalDataAnonymizer    Status: active  Last Run: 2022-11-18 23:41:23 Next Run: 2022-11-18 23:41:23 | Cron |||
2022-11-18 23:41:23 | Debug   | 1     | cron4  3 Tasks planned to run now (2022-11-18 23:41:23): | Cron |||
2022-11-18 23:41:23 | Debug   | 1     | cron4  Task Class: ExecAsyncTask             Status: active  Last Run: 2022-11-18 23:41:21 Next Run: 2022-11-18 23:41:23 | Cron |||
2022-11-18 23:41:23 | Debug   | 1     | cron4  Task Class: CheckStopWatchThresholds  Status: active  Last Run: 2022-11-18 23:41:13 Next Run: 2022-11-18 23:41:23 | Cron |||
2022-11-18 23:41:23 | Debug   | 1     | cron1  >> === 2022-11-18 23:41:23 Start task: ExecAsyncTask ================================== | Cron |||
2022-11-18 23:41:23 | Debug   | 1     | cron4  Task Class: PersonalDataAnonymizer    Status: active  Last Run: 2022-11-18 23:41:23 Next Run: 2022-11-18 23:41:23 | Cron |||
2022-11-18 23:41:23 | Debug   | 1     | cron2  3 Tasks planned to run now (2022-11-18 23:41:23): | Cron |||
2022-11-18 23:41:23 | Debug   | 1     | cron2  Task Class: ExecAsyncTask             Status: active  Last Run: 2022-11-18 23:41:21 Next Run: 2022-11-18 23:41:23 | Cron |||
2022-11-18 23:41:23 | Debug   | 1     | cron2  Task Class: CheckStopWatchThresholds  Status: active  Last Run: 2022-11-18 23:41:13 Next Run: 2022-11-18 23:41:23 | Cron |||
2022-11-18 23:41:23 | Debug   | 1     | cron2  Task Class: PersonalDataAnonymizer    Status: active  Last Run: 2022-11-18 23:41:23 Next Run: 2022-11-18 23:41:23 | Cron |||
2022-11-18 23:41:24 | Debug   | 1     | cron1  ExecAsyncTask: processed 0 tasks | Cron |||
2022-11-18 23:41:24 | Debug   | 1     | cron1  << === 2022-11-18 23:41:24 End of:     ExecAsyncTask ================================== Next: 2022-11-18 23:41:25 | Cron |||
2022-11-18 23:41:24 | Debug   | 1     | cron4  >> === 2022-11-18 23:41:24 Start task: CheckStopWatchThresholds ======================= | Cron |||
2022-11-18 23:41:24 | Debug   | 1     | cron2  >> === 2022-11-18 23:41:24 Start task: ExecAsyncTask ================================== | Cron |||
2022-11-18 23:41:24 | Debug   | 1     | cron1  2 Tasks planned to run now (2022-11-18 23:41:24): | Cron |||
2022-11-18 23:41:24 | Debug   | 1     | cron1  Task Class: PersonalDataAnonymizer    Status: active  Last Run: 2022-11-18 23:41:23 Next Run: 2022-11-18 23:41:23 | Cron |||
2022-11-18 23:41:24 | Debug   | 1     | cron1  Task Class: InlineImageGC             Status: active  Last Run: 2022-11-18 23:41:23 Next Run: 2022-11-18 23:41:24 | Cron |||
2022-11-18 23:41:24 | Debug   | 1     | cron2  ExecAsyncTask: processed 0 tasks | Cron |||
2022-11-18 23:41:24 | Debug   | 1     | cron2  << === 2022-11-18 23:41:24 End of:     ExecAsyncTask ================================== Next: 2022-11-18 23:41:25 | Cron |||
2022-11-18 23:41:24 | Debug   | 1     | cron1  >> === 2022-11-18 23:41:24 Start task: PersonalDataAnonymizer ========================= | Cron |||
2022-11-18 23:41:24 | Debug   | 1     | cron2  1 Tasks planned to run now (2022-11-18 23:41:24): | Cron |||
2022-11-18 23:41:24 | Debug   | 1     | cron2  Task Class: InlineImageGC             Status: active  Last Run: 2022-11-18 23:41:23 Next Run: 2022-11-18 23:41:24 | Cron |||
2022-11-18 23:41:24 | Debug   | 1     | cron2  >> === 2022-11-18 23:41:24 Start task: InlineImageGC ================================== | Cron |||
2022-11-18 23:41:24 | Debug   | 1     | cron2  InlineImageGC: Cleaned 0 old temporary InlineImage(s) and 0 old temporary Attachment(s). | Cron |||
2022-11-18 23:41:24 | Debug   | 1     | cron2  << === 2022-11-18 23:41:24 End of:     InlineImageGC ================================== Next: 2022-11-18 23:41:25 | Cron |||
2022-11-18 23:41:24 | Debug   | 1     | cron2  -- Skipping task: ExecAsyncTask -------------------------- until: 2022-11-18 23:41:25 | Cron |||
2022-11-18 23:41:24 | Debug   | 1     | cron2  -- Skipping task: CheckStopWatchThresholds --------------- until: 2022-11-18 23:41:33 | Cron |||
2022-11-18 23:41:24 | Debug   | 1     | cron2  -- Skipping task: FullTextBackgroundRefresh -------------- until: 2022-11-18 23:41:43 | Cron |||
2022-11-18 23:41:24 | Debug   | 1     | cron2  -- Skipping task: ObsolescenceDateUpdater ---------------- until: 2022-11-18 23:42:07 | Cron |||
2022-11-18 23:41:24 | Debug   | 1     | cron2  -- Skipping task: BulkExportResultGC --------------------- until: 2022-11-19 11:12:51 | Cron |||
2022-11-18 23:41:24 | Debug   | 1     | cron2  -- Skipping task: FullTextBackgroundRebuildFullIndex ----- until: 2022-11-20 01:30:00 | Cron |||
2022-11-18 23:41:24 | Debug   | 1     | cron2  -- Skipping task: LogFileRotationProcess ----------------- until: 2022-12-01 00:00:00 | Cron |||
2022-11-18 23:41:24 | Debug   | 1     | cron2  -- Skipping task: BackupExec ----------------------------- until: 3000-01-01 00:00:00 | Cron |||
2022-11-18 23:41:24 | Debug   | 1     | cron2  sleeping... | Cron |||
```

File: webservices/cron.php (hunk 11 above)
Comment (Hipska) [reply webservices/cron.php:174]: Can be marked as resolved..

File: webservices/cron.php [hunk 15]
```diff
@@ -203,110 +181,121 @@ function RunTask(BackgroundTask $oTask, $iTimeLimit)
...
+			$sCount = count($aDebugMessages);
+			CronLog::Debug("$sCount Tasks planned to run now ($sNow):");
+			foreach ($aDebugMessages as $sDebugMessage) {
+				CronLog::Debug($sDebugMessage);
 			}
 			$aRunTasks = [];
-			foreach ($aTasks as $oTask)
-			{
+			while (count($aTasks) > 0) {
+				$oTask = array_shift($aTasks);
```
Comment (Hipska) [inline webservices/cron.php]: Why this change?
